### PR TITLE
Adapt search from wagtail

### LINKED ISSUE #291: Optional, elasticsearch-powered full-text search
This is not top priority as rolling your own seach is quite easy to do but ideally we want to support this inside the codebase.

Searchly provides a free tier of two indices and 5MB of storage to play with so we can support full text even on Heroku.

### PR DESCRIPTION
**Pull request by artursmet**
Closes #291

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -48,3 +48,6 @@ Environment variables
 
 ``SECRET_KEY``
   Controls `Django's secret key <https://docs.djangoproject.com/en/1.10/ref/settings/#secret-key>`_ setting.
+
+``ELASTICSEARCH_URL``
+  Controls optional `Elasticsearch <https://www.elastic.co/products/elasticsearch>`_ integration. (see :doc:`elasticsearch`) Defaults to ``None``.
```
Comment (patrys) [inline docs/configuration.rst]: "(see ...)" should come before the full stop.

**Comment by patrys** [comment]
We need to either include Wagtail's license with its backend code or ask Torchbox for a separate license. Current license requires its copy to be included with the code.

**Comment by artursmet** [comment]
@patrys I added license from Wagtail.

File: saleor/core/management/commands/populatedb.py [hunk 2]
```diff
@@ -60,3 +70,5 @@ def handle(self, *args, **options):
             else:
                 self.stdout.write(
                     'Superuser already exists - %(email)s' % credentials)
+        if not options['withoutsearch']:
+            self.populate_search_index()
```
Comment (patrys) [inline saleor/core/management/commands/populatedb.py]: Not sure we should call one management from another. I think I'd rather document how various parts play together.

File: saleor/search/backends/dashboard.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+from . import elasticsearch2
```
Comment (patrys) [inline saleor/search/backends/dashboard.py]: This tells me we can't really use any other backends, can we?

Comment (artursmet) [inline saleor/search/backends/dashboard.py]: @patrys Dashboard custom backend is based on elasticsearch2 backend, it should be compatibile with elasticsearch5 as well. DB backend is not supported for dashboard search.

File: saleor/core/management/commands/populatedb.py (hunk 2 above)
Comment (artursmet) [inline saleor/core/management/commands/populatedb.py]: Why not? What's the problem with calling search from this command?

File: saleor/search/backends/dashboard.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            from_=self.start,
+        )
+        params[self.fields_param_name] = 'pk'
+
+        # Add size if set
+        if self.stop is not None:
+            params['size'] = self.stop - self.start
+        # Send to Elasticsearch
+        hits = self.backend.es.search(**params)
+        from collections import defaultdict
```
Comment (patrys) [inline saleor/search/backends/dashboard.py]: Why is this here? `collections` is part of the standard library.

**Comment by artursmet** [comment]
@patrys Unfortunately there is no elasticsearch 5 powered addon on Heroku platform (free)
Searchly uses elasticsearch 2.3.3
Bonsai - 2.4.0

So if we'll support elasticsearch 5 by default - search won't be available on heroku builds.

**Comment by mirekm** [comment]
Surely limiting the ES version only to support the free Heroku account isn't anything we would want to do. Making it clear that demo store does not support search on free Heroku account due to the ES version limitation would be enough IMHO for now.

**Comment by artursmet** [comment]
As @mirekm suggested - I disabled ES integration on default heroku build, so search won't be available there out of the box. Our documentation now contains a note that ES integration requires ES version 5 or higher.
Unused db backend was dropped.

**Comment by patrys** [comment]
Is there a way we could use `import_backend` to automatically base the dashboard backend on the correct implementation? Is the dashboard backend too version-specific?

**Comment by artursmet** [comment]
@patrys I modified the dashboard backend, now it inherits from the base class of active default backend.

File: docs/integrations.rst [hunk 5]
```diff
@@ -104,3 +104,15 @@ To get it working you need to export the following environment variable:
...
   Your page's Google “Tracking ID.“
+
+
+Full text search with Elasticsearch
+-----------------------------------
+
+You can use optional `Elasticsearch <https://www.elastic.co/products/elasticsearch>`_ integration.
+To get it working you need to export following environment variable:
+
+``ELASTICSEARCH_URL``
```
Comment (patrys) [inline docs/integrations.rst]: I'd drop the variable from here as there's more than one and there is no point in duplicating the descriptions.

File: docker-compose.yml [hunk 6]
```diff
@@ -11,18 +11,24 @@ services:
...
+    image: elasticsearch:5.1.1
+    ports:
+      - '9200:9200'
   web:
     build: .
     command: python manage.py runserver 0.0.0.0:8000
     environment:
       - DATABASE_URL=postgres://saleor:saleor@db/saleor
       - DEFAULT_FROM_EMAIL=noreply@example.com
+      - ELASTICSEARCH_URL=http://search:9200/
```
Comment (patrys) [inline docker-compose.yml]: `ELASTICSEARCH_API=5`?

Comment (artursmet) [inline docker-compose.yml]: As discussed, we'll use ES 2.X by default.
